crates/sui-core/src/consensus_manager/mod.rs

```diff
@@ -57,8 +68,153 @@ impl ConsensusManager {
+                "The number of retries took to start narwhal primary node",
+                registry
+            )
+            .unwrap(),
+            start_worker_retries: register_int_gauge_with_registry!(
+                "narwhal_manager_start_worker_retries",
+                "The number of retries took to start narwhal worker node",
+                registry
+            )
+            .unwrap(),
+            start_mysticeti_retries: register_int_gauge_with_registry!(
+                "mysticeti_manager_start_retries",
+                "The number of retries took to start Mysticeti node",
+                registry
+            )
+            .unwrap(),
+        }
+    }
+}
+
+pub(crate) struct RunningLockGuard<'a> {
+    state_guard: MutexGuard<'a, Running>,
+    metrics: &'a ConsensusManagerMetrics,
+    completed: bool,
+    epoch: Option<EpochId>,
+    protocol_version: Option<ProtocolVersion>,
+    start: Instant,
+}
+
+impl<'a> RunningLockGuard<'a> {
+    pub(crate) async fn acquire_start(
+        metrics: &'a ConsensusManagerMetrics,
+        running_mutex: &'a Mutex<Running>,
+        epoch: EpochId,
+        version: ProtocolVersion,
+    ) -> Option<RunningLockGuard<'a>> {
+        let running = running_mutex.lock().await;
+        if let Running::True(epoch, version) = *running {
+            tracing::warn!(
+                "Consensus is already Running for epoch {epoch:?} & protocol version {version:?} - shutdown first before starting",
+            );
+            return None;
+        }
+
+        tracing::info!("Starting up consensus for epoch {epoch:?} & protocol version {version:?}");
+
+        Some(RunningLockGuard {
+            state_guard: running,
+            metrics,
+            completed: false,
+            start: Instant::now(),
+            epoch: Some(epoch),
+            protocol_version: Some(version),
+        })
+    }
+
+    pub(crate) async fn acquire_shutdown(
+        metrics: &'a ConsensusManagerMetrics,
+        running_mutex: &'a Mutex<Running>,
+    ) -> Option<RunningLockGuard<'a>> {
+        let running = running_mutex.lock().await;
+        if let Running::True(epoch, version) = *running {
+            tracing::info!(
+                "Shutting down consensus for epoch {epoch:?} & protocol version {version:?}"
+            );
+        } else {
+            tracing::warn!("Consensus shutdown was called but Narwhal node is not running");
+            return None;
+        }
+
+        Some(RunningLockGuard {
+            state_guard: running,
+            metrics,
+            completed: false,
+            start: Instant::now(),
+            epoch: None,
+            protocol_version: None,
+        })
+    }
+
+    // has to be called when the process of start/shutdown has been finished in order to mark correctly timings.
+    fn completed(&mut self) {
+        self.completed = true;
+    }
+}
+
+impl Drop for RunningLockGuard<'_> {
+    fn drop(&mut self) {
+        if self.completed {
```

Comment:
This might be unnecessary as once the guard has been acquired the start/shutdown methods on the managers will either succeed or panic - but to avoid any accidental errors I chose to explicitly mark them as completed